File: components/animation/src/main/java/org/storm/animations/sprite/SpriteAnimation.java
package org.storm.animations.sprite;

import javafx.scene.canvas.GraphicsContext;
import javafx.scene.image.Image;
import lombok.Getter;
import org.storm.animations.Animation;

/**
 * A specific type of Animation focusing on the use of 2D Sprites and SpriteSheets.
 */
public class SpriteAnimation implements Animation {

  private final Image[] sprites;

  private final int loopCount;

  private final int frameDelay;

  private int currentFrame;

  private int currentLoop;

  private int currentSprite;

  @Getter
  private boolean playing;

  public SpriteAnimation(Image[] sprites, int frameDelay, int loopCount) {
    if (loopCount != LOOP_INDEFINITELY && loopCount < 0) {
      throw new IllegalArgumentException("loopCount must be >= 0");
    }

    this.sprites = sprites;
    this.frameDelay = frameDelay;
    this.loopCount = loopCount;
    this.playing = true;
  }

  @Override
  public void play() {
    this.playing = true;
  }

  @Override
  public void pause() {
    this.playing = false;
  }

  public void reset() {
    this.currentFrame = 0;
    this.currentSprite = 0;
    this.currentLoop = 0;
  }

  @Override
  public void render(GraphicsContext graphicsContext, double x, double y) {
    graphicsContext.drawImage(this.sprites[this.currentSprite], x, y);
  }

  @Override
  public void update(double time, double elapsedTime) {

    if (!this.playing || this.currentLoop == this.loopCount) return;

    this.currentFrame++;

    if (this.currentFrame == this.frameDelay) {
      this.currentFrame = 0;
      this.currentSprite++;
    }

    if (this.currentSprite == this.sprites.length) {
      this.currentSprite = 0;
      this.currentLoop = (this.currentLoop == LOOP_INDEFINITELY) ? LOOP_INDEFINITELY : this.currentLoop + 1;
    }

  }

}

File: components/animation/src/main/kotlin/org/storm/animations/sprite/SpriteAnimation.kt
package org.storm.animations.sprite

import org.storm.animations.Animation
import org.storm.core.asset.Asset
import org.storm.core.render.canvas.Canvas
import org.storm.core.render.Image

/**
 * A specific type of Animation focusing on the use of 2D Sprites and SpriteSheets.
 */
@Asset("animation", "sprite")
class SpriteAnimation(
    override val delay: Int,
    override val loops: Int,
    private val sprites: Array<Image>
) : Animation {

    private var currentFrame = 0
    private var currentLoop = 0
    private var currentSprite = 0
    private var hasCompletedOnce = false
    private var playing: Boolean = true

    init {
        require(!(loops != Animation.LOOP_INDEFINITELY && loops < 0)) {
            "loopCount must be >= 0"
        }
    }

    override fun play() {
        playing = true
    }

    override fun pause() {
        playing = false
    }

    override fun reset() {
        currentFrame = 0
        currentSprite = 0
        currentLoop = 0
    }

    override fun isPlaying(): Boolean {
        return playing
    }

    override fun isComplete(): Boolean {
        return (loops == Animation.LOOP_INDEFINITELY && hasCompletedOnce) || currentLoop == loops
    }

    override suspend fun render(canvas: Canvas, x: Double, y: Double) {
        canvas.drawImageWithPixels(sprites[currentSprite], x, y)
    }

    override suspend fun update(time: Double, elapsedTime: Double) {
        if (!playing || currentLoop == loops) return

        if (++currentFrame == delay) {
            currentFrame = 0
            currentSprite++
        }

        if (currentSprite == sprites.size) {
            currentSprite = 0
            currentLoop = if (currentLoop == Animation.LOOP_INDEFINITELY) Animation.LOOP_INDEFINITELY else currentLoop + 1
            hasCompletedOnce = true
        }
    }
}
